core/psh/nc.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Phoenix-RTOS
+ *
+ * nc - TCP and UDP connect or listen
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <sys/select.h>
+
+#include "psh.h"
+
+
+/*
+ * This could be achieved using getaddrinfo(),
+ * however it does not recognize scope id for now
+ * and requires dynamic memory, which we
+ * are trying to avoid.
+ */
+static int psh_nc_sockaddrFill(int *af_family, char *addrstr, char *portstr,
+                               struct sockaddr_storage *sockaddr,
+                               socklen_t *addrlen)
+{
+	struct sockaddr_in6 *saddr_in6;
+	struct sockaddr_in *saddr;
+	char *ifname, *endptr;
+	int port = 0, scopeid = 0;
+	uint32_t addr[4] = { };
+
+	if (*af_family == AF_UNSPEC) {
+		if (addrstr && strchr(addrstr, ':') != NULL)
+			*af_family = AF_INET6;
+		else
+			*af_family = AF_INET;
+	}
+
+	/* Try to resolve scope id */
+	if (*af_family == AF_INET6) {
+		ifname = strtok(addrstr, "%");
```

Comment:
```
ifname = strtok(addrstr, "%");
```
What if:
- `addrstr == NULL`
-  `ifname == NULL`

Reply:
I agree, we should add a check: `addrstr != NULL`. 
However, `strtok` should return `NULL` when calling it again even if it returned `NULL` before.

Reply:
Look at the code for the `strtok()` function [libphoenix/blob/master/string/string.c](https://github.com/phoenix-rtos/libphoenix/blob/6e57d09a03ed8c30af3751abf4c36fb9bc8af59a/string/string.c#L369)

Reply:
I understand, we are considering a case, when a first call to `strtok()` returns `NULL`, so if a token `%` is not found. `strtok()` sets is static variable `string_common.next_token` to point to an empty string `"\0"`.
When we are calling this function again, one line below, the `strtok()` would again return NULL, so everything is fine there.
I already agreed with you, that cheking `addrstr` is necessary, but I don't know, what you mean with this `ifname` variable.

Reply:
Ok. It is enough to check `addrstr`.